plutus-core/plutus-core/src/PlutusCore/Builtin/Runtime.hs

```diff
@@ -51,7 +54,12 @@ instance NFData (RuntimeScheme n) where
 -- argument of the denotation and calling 'makeKnown' over its result.
 type ToRuntimeDenotationType :: GHC.Type -> Peano -> GHC.Type
 type family ToRuntimeDenotationType val n where
-    ToRuntimeDenotationType val 'Z     = MakeKnownM val
+    -- We use 'Lazy' here, because we don't want to compute the denotation when it's fully saturated
+    -- before figuring out what it's going to cost. An alternative is to be really careful and never
```

Comment:
> we don't want to compute the denotation when it's fully saturated before figuring out what it's going to cost

What does this mean? Do we sometimes not compute the result if the cost is too high?

Reply:
> Do we sometimes not compute the result if the cost is too high?

Exactly. If something costs more than what the budget allows, we don't compute the builtin application and return an out-of-ex failure immediately.